File: app/src/main/java/com/b_lam/resplash/util/ThemeUtils.java
package com.b_lam.resplash.util;

import android.app.Activity;
import android.app.ActivityManager;
import android.content.Context;
import android.content.SharedPreferences;
import android.content.res.TypedArray;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Build;
import android.preference.PreferenceManager;
import androidx.annotation.AttrRes;
import androidx.annotation.ColorInt;
import androidx.annotation.DrawableRes;
import androidx.annotation.IntDef;
import androidx.annotation.StringDef;

import com.b_lam.resplash.R;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

/**
 * Created by Brandon on 6/13/2017.
 */

public class ThemeUtils {

    private final static String TAG = "ThemeUtils";

    public @StringDef({Theme.LIGHT, Theme.DARK, Theme.BLACK})
    @Retention(RetentionPolicy.SOURCE)
    @interface Theme {
        String LIGHT = "Light";
        String DARK = "Dark";
        String BLACK = "Black";
    }

    public static @Theme String getTheme(Context context) {
        SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context);
        return sharedPreferences.getString("theme", Theme.LIGHT);
    }

    @ColorInt
    public static int getThemeAttrColor(Context context, @AttrRes int colorAttr) {
        TypedArray array = context.obtainStyledAttributes(null, new int[]{colorAttr});
        try {
            return array.getColor(0, 0);
        } finally {
            array.recycle();
        }
    }

    @DrawableRes
    public static int getThemeAttrDrawable(Context context, @AttrRes int drawableAttr) {
        TypedArray array = context.obtainStyledAttributes(null, new int[]{drawableAttr});
        try {
            return array.getResourceId(0, 0);
        } finally {
            array.recycle();
        }
    }

    public static void setRecentAppsHeaderColor(Activity activity) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            Bitmap icon = BitmapFactory.decodeResource(activity.getResources(), R.mipmap.ic_launcher);
            ActivityManager.TaskDescription taskDescription
                    = new ActivityManager.TaskDescription(
                    activity.getString(R.string.app_name),
                    icon, getThemeAttrColor(activity, R.attr.colorPrimary));
            activity.setTaskDescription(taskDescription);
            if (icon != null) icon.recycle();
        }
    }
}

File: app/src/main/java/com/b_lam/resplash/util/ThemeUtils.kt
package com.b_lam.resplash.util

import android.content.Context
import android.content.res.TypedArray
import androidx.annotation.AttrRes
import androidx.annotation.ColorInt
import androidx.appcompat.app.AppCompatDelegate
import androidx.browser.customtabs.CustomTabsIntent.*

private const val LIGHT = "light"
private const val DARK = "dark"
private const val BATTERY = "battery"
private const val DEFAULT = "default"

fun applyTheme(theme: String?) {
    when (theme) {
        LIGHT -> AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO)
        DARK -> AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES)
        BATTERY -> AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_AUTO_BATTERY)
        DEFAULT -> AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM)
    }
}

@ColorScheme
fun getCustomTabsColorScheme(theme: String?): Int {
    return when (theme) {
        LIGHT -> COLOR_SCHEME_LIGHT
        DARK -> COLOR_SCHEME_DARK
        else -> COLOR_SCHEME_SYSTEM
    }
}

@ColorInt
fun getThemeAttrColor(context: Context, @AttrRes colorAttr: Int): Int {
    val array: TypedArray = context.obtainStyledAttributes(null, intArrayOf(colorAttr))
    return try {
        array.getColor(0, 0)
    } finally {
        array.recycle()
    }
}
